// translate/index.js
'use strict';

const https=require('https');
const http=require('http');

function naver(q,tl){
  // q: query string
  // tl: target language
  return new Promise(function (resolve,reject){
    if(typeof q!=='string' || q.trim().length===0){
      resolve([]);
    }
    q=encodeURIComponent(q);

    if(typeof tl!=='string') tl='cn';
    tl=tl.toLowerCase();

    let req, _p,
    _y=function(res){
      let s=[],i=0;
      
      // console.debug(res.statusCode);
      if(res.statusCode!==200) resolve([]);

      res.on('data',d=>{
        i+=d.length;
        s.push(d);
      });
      res.on('end',()=>{
        let r=Buffer.concat(s,i).toString('utf8');
        r=_p(r);
        // console.debug(r)
        resolve(r);
      });
    },
    _n=function(v){y([])},
    _p1=function(s){
      let j=JSON.parse(s);
      let r=[];

      // console.debug(j.items.map(e=>e.map(e=>e.map(e=>e[0]))));

      j.items[0].forEach(e=>{
        e=e.map(e=>e.join());
        if(e[0]==='2'){
          r.push({kr:e[1],'zh-CN':e[2]});
        }else if(e[0]==='1'){
          r.push({kr:e[5],cn:e[1]});
        }
      });
      return r;
    },
    _p2=function(s){
      // console.debug(s);// 
      let j=JSON.parse(s);
      let r=[];
      // console.debug(j.items.map(e=>e.map(e=>e[0])));
      j.items.forEach(e=>{
        let a=e.map(e=>e[0]);
        r=r.concat(a);
      });
      return r;
    },
    _p3=function(s){
      // console.debug(s);
      let j=JSON.parse(s);
      let r=j.items.map(e=>{
        let a=e.split('|').slice(1);
        return {cn:a[0],ko:a[1]};
      });
      return r;
    },
    _p4=function(s){
      // console.debug(s);// items:[ [[s],[t]] ]
      let j=JSON.parse(s);
      let r=[];
      j.items.map(e=>e.map(e=>{
        let a=e.map(e=>e[0]);
        // console.debug('---', a)
        r.push({kr:a[0],jp:a[1]});
      }));
      return r;
    },
    _p5=function(s){
      console.debug(s.toString());
    };

    if(tl==='cn') _p=_p1, req=https.get('https://ac.dict.naver.com/zhdict/ac?mod=json&st=11&r_lt=10&q='+q, _y).on('error',_n);
    if(tl==='kr')  _p=_p2, req=https.get('https://ac.dict.naver.com/krdic/ac?frm=stdkrdic&oe=utf8&m=0&r=1&st=111&r_lt=111&q='+q, _y).on('error',_n);
    if(tl==='ch') _p=_p3, req=http.get('http://suggest-bar.daum.net/suggest?mod=json&code=utf_in_out&enc=utf&id=language&cate=han&q='+q, _y).on('error',_n);
    if(tl==='jp') _p=_p4, req=https.get('https://ja.dict.naver.com/ac?st=111&r_lt=111&n_kojpdic=1&q='+q, _y).on('error',_n);
  });
};


// test1: q===undefined
// naver().then(e=>console.log(e))
// naver(null).then(e=>console.log(e))

// naver('하나','kr').then(e=>console.log(e));// [ko,...]

// 汉字
// naver('일','ch').then(e=>console.log(e));// [[cn,ko],...]
// naver('一','ch').then(e=>console.log(e));// [[cn,ko],...]

// naver('일','jp').then(e=>console.log(e));

// naver('中文','cn').then(e=>console.log(e));



function google(q,tl){
  return new Promise((resolve,reject)=>{
    if(typeof q!=='string' || q.trim().length===0) resolve([]);
    q=encodeURIComponent(q);

    let url='http://translate.google.cn/translate_a/single?client=gtx&dt=t&ie=UTF-8&oe=UTF-8&sl=auto&tl='+tl+'&q='+q;
    http.get(url, function(res){
      if(res.statusCode!==200) resolve([]);
      let s=[],i=0;
      res.on('data',d=>{
        i+=d.length;
        s.push(d);
      });
      res.on('end',()=>{
        let r=Buffer.concat(s,i).toString('utf8');
        let j=JSON.parse(r);
        // console.debug(r)
        resolve({ [j[2]] : j[0].map(e=>e[0]).join() })
      });
    }).on('error',e=>resolve([]));
  });
}

function googles(q,tl='zh-CN'){
  function vq(a,uq="426928.3386708673") {
      if (null !== uq)
          var b = uq;
      else {
          b = sq('T');
          var c = sq('K');
          b = [b(), c()];
          b = (uq = window[b.join(c())] || "") || ""
      }
      var d = sq('t');
      c = sq('k');
      d = [d(), c()];
      c = "&" + d.join("") + "=";
      d = b.split(".");
      b = Number(d[0]) || 0;
      for (var e = [], f = 0, g = 0; g < a.length; g++) {
          var l = a.charCodeAt(g);
          128 > l ? e[f++] = l : (2048 > l ? e[f++] = l >> 6 | 192 : (55296 == (l & 64512) && g + 1 < a.length && 56320 == (a.charCodeAt(g + 1) & 64512) ? (l = 65536 + ((l & 1023) << 10) + (a.charCodeAt(++g) & 1023),
          e[f++] = l >> 18 | 240,
          e[f++] = l >> 12 & 63 | 128) : e[f++] = l >> 12 | 224,
          e[f++] = l >> 6 & 63 | 128),
          e[f++] = l & 63 | 128)
      }
      a = b;
      for (f = 0; f < e.length; f++)
          a += e[f],
          a = tq(a, "+-a^+6");
      a = tq(a, "+-3^+b+-f");
      a ^= Number(d[1]) || 0;
      0 > a && (a = (a & 2147483647) + 2147483648);
      a %= 1000000;
      return c + (a.toString() + "." + (a ^ b))
  };
  function sq(a) {
      return function() {
          return a
      }
  }
  function tq(a, b) {
      for (var c = 0; c < b.length - 2; c += 3) {
          var d = b.charAt(c + 2);
          d = "a" <= d ? d.charCodeAt(0) - 87 : Number(d);
          d = "+" == b.charAt(c + 1) ? a >>> d : a << d;
          a = "+" == b.charAt(c) ? a + d & 4294967295 : a ^ d
      }
      return a
  }
  return new Promise((resolve,reject)=>{
    if(typeof q!=='string' || q.trim().length===0) resolve([]);
    let tk=vq(q);
    q='&q='+encodeURIComponent(q);
    let url='https://translate.google.cn/translate_a/single?client=t&sl=auto&tl='+tl+'&hl=zh-CN&dt=at&dt=bd&dt=ex&dt=ld&dt=md&dt=qca&dt=rw&dt=rm&dt=ss&dt=t&ie=UTF-8&oe=UTF-8&otf=2&ssel=0&tsel=0&kc=1'+tk+q;
    https.get(url, res=>{
      if(res.statusCode!==200) resolve([]);
      let s=[],i=0;
      res.on('data',d=>{
        i+=d.length;
        s.push(d);
      });
      res.on('end',()=>{
        let r=Buffer.concat(s,i).toString('utf8');
        let j=JSON.parse(r);
        resolve({ [j[2]] : j[0]&&j[0].map(e=>e[0]).join('') });
      });
    }).on('error', ()=>resolve([]));
  });
};




// google('你好\n早安','en')
// google('你好\n早安','en').then(e=>console.log(e));// { t: 'Hello there\ngood Morning', sl: 'zh-CN' }
// google('你好\n早安','ja').then(e=>console.log(e));// { t: 'ハロー\nおはよう', sl: 'zh-CN' }
// google('你好\n早安','ko').then(e=>console.log(e));// { t: '안녕하세요.\n안녕하세요.', sl: 'zh-CN' }
// google('你好\n早安','zh-TW').then(e=>console.log(e));// { t: '你好\n早安', sl: 'zh-CN' }
// google('set','zh-cn').then(e=>console.log(e));// { t: '你好\n早安', sl: 'zh-CN' }
// googles('set','ko').then(e=>console.log(e));// { t: '你好\n早安', sl: 'zh-CN' }


function daum(q,tl) {
  function lang(l) {
    if(typeof l==='string') l=l.toLowerCase();
    switch(l) {
      case 'en':
      case 'eng':
      case 'english':
        return 'eng'
      case 'jp':
      case 'jpn':
      case 'japan':
        return 'jpn';
      case 'chn':
      case 'chinese':
      case 'china':
      case 'zh-cn':
      case 'cn':
      case 'ch':
        return 'chn';
      case 'hanja':
        return 'hanja';
      case 'fr':
      case 'french':
        return 'fr';
      case 'ko':
      case 'kor':
      case 'korea':
      default:
        return 'kor';
    }
  }
  return new Promise((resolve,reject)=>{
    if(!q) return resolve(null);
    let url='http://suggest-bar.daum.net/suggest?mod=json&code=utf_in_out&enc=utf&id=language&cate='+lang(tl)+'&q='+encodeURIComponent(q);
    http.get(url, function(res){
      if(res.statusCode!==200) resolve([]);
      let s=[],i=0;
      res.on('data',d=>{
        i+=d.length;
        s.push(d);
      });
      res.on('end',()=>{
        let r=Buffer.concat(s,i).toString('utf8');
        let j=JSON.parse(r);
        // j.slice(1).forEach(e=>console.log('---',e));
        r=j.items.map(e=>{
          let row=e.split('|');
          return {s:row[1],t:row[2]};
        });
        r=[] .concat(...r);
        resolve(r)
      });  
    });
  });
}


// daum(`그림`,'en').then(d=>console.log(d));// [{s,t},...]
// daum(`그림`,'jpn').then(d=>console.log(d));// [{s,t},...]
// daum(`그림`,'chn').then(d=>console.log(d));// [{s,t},...]
// daum(`그림`,'fr').then(d=>console.log(d));// [{s,t},...]
// daum(`그림`,'ko').then(d=>console.log(d));// [{s,t},...]


function hanja(q) {
  return new Promise((resolve,reject)=>{
    if(!q) return resolve(null);
    let url='http://alldic.daum.net/search.do?mod=json&dic=hanja&q='+encodeURIComponent(q);
    http.get(url, function(res){
      if(res.statusCode!==200) resolve([]);
      let s=[],i=0;
      res.on('data',d=>{
        i+=d.length;
        s.push(d);
      });
      res.on('end',()=>{
        let r=Buffer.concat(s,i).toString('utf8');
        let cheerio=require('cheerio');
        let $=cheerio.load(r);

        r=[];

        {
          let ja=$.root().find('.hanja_box');
          let res={
            hanja:[],
            example:[]
          };

          ja.find('.search_word').each((i,e)=>{
            let hanja=$(e).text().replace(/([\t\n ])+/g,' ').trim();
            res.hanja[i]=[];
            res.hanja[i].push(hanja.slice(0,1), hanja.slice(2))
          });
          // res.hanjas=hanjas;

          ja.find('.list_search').each(function(i,e){
            let spans=[];
            $(e).find('li').each((i,li)=>{
              let span=$(li).find('span:nth-child(2)').text();
              spans.push(span);
            });
            res.hanja[i].push(...spans);
          });

          $.root().find('.kokk_type').each((i,e)=>{
            let k=$(e).find('.tit_searchword a').map((i,e)=>$(e).text()).toArray();
            let d=$(e).find('.txt_search').text().trim()
            k.push(d);
            res.example.push(k)
          });
          resolve(res)
        }
      });  
    });
  });
}

// hanja(`어`) .then(d=>console.log(d));// [{s,t},...]


module.exports = {
  google,
  googles,
  googleLanguages:['en','ja','ko','zh-CN','zh-TW'],
  naver,
  naverLanguages:['cn','ch','jp','kr'],
  daum,
  daumLanguages:['eng','jpn','chn','kor','fr'],
  hanja
}
